README.MD
# gst-pylonsrc

This is a gstreamer element that allows using Basler's USB3 Vision cameras using gstreamer.

This package also includes a simple plugin to check the framerate of any given pipeline called `fpsfilter`.

These plugins were open sourced by PlayGineering Ltd. on June 12, 2018.

## Requirements
#### To run
* Linux (Tested with Ubuntu 14.04 and Ubuntu 16.04)
* [Gstreamer-1.0 (Tested with gstreamer 1.8.3)](https://gstreamer.freedesktop.org/src/)
    * For older cameras You'll need the bayer2rgb plugin available in gst-plugins bad.
    * To record to file you'll need an encoder (i.e. x264enc).
* [Pylon 5.0.5 or newer (Tested with Pylon 5.0.5)](http://www.baslerweb.com/en/support/downloads/software-downloads)

#### To compile
* All of the above plus
* gcc 4.8 or newer (Tested with gcc 4.8.5). Note that this plugin will only compile with gcc or clang.
* GNU Make 3.8 or newer (Tested with gnu make 3.8.1)
* autoconf 2.69 or newer (Tested with autoconf 2.69)
* automake 1.14 or newer (Tested with automake 1.14.1)
* libtool 2.4 or newer (Tested with libtool 2.4.2)

## Compiling
After cloning the repository run the `autogen.sh` file in the root directory of the project. Autoconf will generate everything required for compilation. After it completes simply run the `make` command in the root directory and it should compile.

## Installation
After compiling there are two ways to install this plugin - automated and manual.

To install automatically simply run `sudo make install` inside the root directory of this project and Automake will automatically install the plugin.

To install manually you need copy, move or symlink the following files

* `./plugins/libgstpylonsrc.la`
* `./plugins/.libs/libgstpylonsrc.so`

to either of these locations:

* `~/.local/share/gstreamer-1.0/plugins/` if you only want your user to be able to use this plugin
* `/usr/local/lib/gstreamer-1.0/` if you want this plugin to be available to every user.

Note that these are the default locations which might be different if you installed gstreamer with a custom location prefix.

## Usage
To test whether this plugin was installed correctly use the `gst-inspect-1.0 pylonsrc` command. If it returns information about the plugin, then it's installed successfully and can be used like any other gstreamer source.

An example pipeline with this plugin for an older camera with no rgb8 support would be `gst-launch-1.0 -v pylonsrc ! bayer2rgb ! videoconvert ! xvimagesink`.

You set the parameters the same way one would set parameters for any other gstreamer pipeline element - by appending `key=value` to the plugin. For example, to get ycbcr422_8 output from the camera, you would write your pipeline like this - `gst-launch-1.0 pylonsrc imageformat=rgb8 ! videoconvert ! xvimagesink`.

#### Formats
This plugin supports the following output formats:
 + Bayer 8 (default) (value: `bayer8`)
 + Bayer 10 (value: `bayer10`)
 + Bayer 10 packed (value: `bayer10p`)
 + RGB8 (value: `rgb8`)
 + BGR8 (value: `bgr8`)
 + Mono8 (Greyscale) (value: `mono8`)
 + YcbCr422 8 (value: `ycbcr422_8`)

To change the output format simply set the `imageformat` parameter to the value of one of the formats above. Note that with the bayer formats you want to use the `bayer2rgb` plugin available in `gst-plugins-bad` (followed by `videoconvert` if you want the stream in YUY2 or similar format). RGB, BGR and Mono8 streams might require `videoconvert` for output to screen (video sinks such as `xvimagesink` only seems to play nice with YUY2 despite claiming to work with RGB).

#### Properties
If you have multiple cameras you can specify the camera you want to use using the `camera` parameter. To find out the IDs for each camera simply launch the pipeline without specifying the `camera` parameter, and the plugin will output them.

Each camera has a custom field that is saved on the camera called `userid`. This parameter allows user to quickly identify cameras on the field. This parameter will be listed when connecting to a camera or listing the devices. It can be set by either passing the `userid` parameter to the plugin, or by setting it in the Pylon Viewer App (`camera -> Device Control -> Device User ID`).

If you want to set a custom resolution you can do so with the `width` and `height` parameters. To find out the maximum resolution of the camera either refer to camera's technical specs or simply launch the plugin with logging tuned to loglevel4 (GST_DEBUG=pylonsrc:4), and it will output the maximum resolution the camera is capable of. Please note that the camera actually saves the resolution you set, so if you launch a pipeline with a camera once while specifying a resolution then the following runs will use the same resolution unless you either reconnect the device or specify a different resolution.

You can also specify resolution offsets to slightly move the image. Possible offsets are calculated `max(width|height)-current(width|height)`. The parameters to specify offsets are `offsetx` and `offsety`. If you want to center the image you can use `centerx` and `centery`. Note that setting the centering parameters will cause the plugin to ignore the offset values. 

The `limitbandwidth` parameter can be used to disable bandwidth limitations on the cameras. Basler cameras seem to limit themselves to 300MB/s (which is approx 130fps with a 1920x1200 resolution). Depending on the camera, it might be required to change the sensor readout mode (`sensorreadoutmode`) to fast for this to have any effect, as the default settings are usually close to the normal readout mode's maximum values. The possible values for this are either `on` or `off`. If you want to limit bandwidth to a specific number then set this property to `on` (or just don't specify it as `on` is the default) and use the `maxbandwidth` parameter which takes in the number of bytes per second that the camera will send data at. Note that setting the value too low will cause the camera to send data at a lower framerate than it is capable of, while setting it too high will in most cases have no discernible effect as the camera will only use as much bandwidth as it needs for the specified resolution and framerate. To see the bandwidth use run the pipeline with debugging level set to 4 (i.e. by adding `GST_DEBUG=pylonsrc:4` before the gst-launch-1.0 command) and all of the information related to framerate and bandwidth will be printed to screen during the initialisation procedure.

To achieve the full potential of the camera it might be required to set the sensor readout mode to fast. To do this use the `sensorreadoutmode` parameter which takes in either `fast` or `normal` as values. As the name implies, `fast` allows for higher framerates, while `normal` is the safe (and default) option.

To change the framerate use the `fps` property.

To have the camera send a test pattern instead of a video stream you can use the `testimage` property. It accepts values from 1-6 all of which correspond to different test patterns.

There are two modes for picture capture - trigger mode and continuous. Trigger mode asks for each frame seperately which while continuous mode makes the camera capture frames without software input. Continuous is default, but it can be disabled by setting the `continuous` parameter to `false` (default - `true`).

NOTE: Some of the parameters are saved to the camera. Running the pipeline multiple times without either reconnecting the device or using the `reset` parameter might cause weird behaviour. See the `gst-inspect-1.0` output for more details.

#### Image settings
To fine tune the image you can use the following properties:

* `lightsource` - Available values are `off`, `2800k`, (default)`5000k` and `6500k`. Changes the colour temperature of the video to have the best visibility in the given environment.
* `autoexposure` - Available values are (default)`off`, `once` and `continuous`. Enable automatic exposure. Overrides the `exposure` setting.
* `exposurelowerlimit`, `exposureupperlimit` - Sets the limit for auto exposure function. Values range from `105.0` to `1000000.0`.
* `exposure` - Accepts a number (microseconds). Sets the exposure time for each frame. `autoexposure` must be set to `off` for this to have any effect.
* `autowhitebalance` - Available values are `off`, `once` and `continuous`. Determines whether or not the camera will try to determine the best colour balance settings.
* `balancered`, `balancegreen` and `balanceblue` - Accepts a floating point number from `0` to `15,9`. Changes the colour balance for the specific colour. Defaults to whatever is set on the camera.
* `colorredhue`, `coloryellowhue`, `colorgreenhue`, `colorcyanhue`, `colorbluehue` and `colormagentahue` - Accepts a floating point number from -4 to 3.9. Changes the hue for the specific colour. Defaults to whatever is set on the camera.
* `colorredsaturation`, `coloryellowsaturation`, `colorgreensaturation`, `colorcyansaturation`, `colorbluesaturation` and `colormagentasaturation` - Accepts a floating point number from 0 to 1.9. Changes the saturation for the specific colour. Defaults to whatever is set on the camera.
* `transformationselector` - Sets the type of colour transformation used. This is used when using the `transformation` parameters. Available values are `rgbrgb`, `rgbyuv` and `yuvrgb`. The first colour is the origin format, the second format is the format that the colour will be transformed to. 
* `transformation00`, `transformation01`, `transformation02`, `transformation10`, `transformation11`, `transformation12`, `transformation20`, `transformation21`, `transformation22` allows for colour transformations within each colour channel. It's recommended to play around with the sliders in PylonViewerApp to find the values desired.
* `autogain` - Available values are `off`, `once` and `continuous`. Determines whether or not the camera will automatically adjust the gain option.
* `gainupperlimit`, `gainlowerlimit` - Sets the limit for auto gain function. Values range from `0` to `12.00921`.
* `gain` - Accepts a floating point number (dB) from `0` to `12`. Changes the amount of gain that will be added to the picture before it will be sent to the computer. `autogain` must be set to `off` for this to have any effect.
* `autobrightnesstarget` - Sets the target value for auto exposure function. Values range between `0.19608` and `0.80392`.
* `blacklevel` - Accepts a floating point number (DN) from `0` to `63.75`. Determines colour black's level.
* `gamma` - Accepts values from `0` to `3.9`. Changes the gamma of the picture.
* `sharpnessenhancement` - Uses basler's post-processing system to apply sharpness enhancement. This will only work if the `imageformat` is not bayer*. Values range from `1.0` to `3.98`. `demosaicing` must be set to true first.
* `noisereduction` - Uses basler's post-processing system to apply noise reduction. This will only work if the `imageformat` is not bayer*. Values range from `0` to `2.0`. `demosaicing` must be set to true fist.
* `autoprofile` - Determines whether the automatic brightness etc. functions will prioritise minimising gain or exposure. Values are `gain` and `exposure`.
#### Example pipelines
Output to screen: `GST_DEBUG=pylonsrc:5 gst-launch-1.0 pylonsrc imageformat=ycbcr422_8 ! xvimagesink`

Recording: `GST_DEBUG=pylonsrc:5 gst-launch-1.0 pylonsrc camera=0 ! bayer2rgb ! videoconvert ! x264enc noise-reduction=10000 pass=4 quantizer=22 bitrate=32768 speed-preset=0 byte-stream=true ! matroskamux ! filesink location='recording.mkv' sync=false`

Record a video at 150fps: `GST_DEBUG=pylonsrc:5 gst-launch-1.0 pylonsrc limitbandwidth=off sensorreadoutmode=fast fps=150 ! bayer2rgb ! videoconvert ! matroskamux ! filesink location='recording.mkv`

## fpsfilter
This package includes a simple plugin called `fpsfilter`. To use it simply plug it in any pipeline you want.

For example - `GST_DEBUG=pylonsrc:5 gst-launch-1.0 pylonsrc ! fpsfilter ! bayer2rgb ! videoconvert ! xvimagesink`.

The output is displayed in the console. You can control the frequency of the reports using `reporttime` parameter, which (in miliseconds) defines the time between messages that are sent to the screen. (default - 1000ms).

## Misc
If you need to reset the camera(s) quickly but don't want to reset it(them) using the reset parameter, you can use the `reset.sh` file in the tools directory which will reset the USB devices.

## TODO

* Rather than set every parameter ourselves, first get the value from camera and use that one if applicable.
* Convert all string literals to gstrings.


CHANGELOG
2018-06-12 Ingmars Melkis <contact@zingmars.me>

  * Open sourced the plugin with CEO's permission.

2017-05-18 Ingmars Melkis <zingmars@playgineering.com>

  * Added a FPS counter plugin.
  * Implemented all of the output types our cameras support.
  * Implemented continuous grabbing support.
  * Added colour transformation configuration.
  * Added AutoExposure upper and lower limits.
  * Added a configuration option for automatic brightness.
  * Added resolution offsets.
  * Added support for PGI.
  * Added support for image flipping.
  * Implemented multiple camera support.
  * Added more debug output.
  * Added configure checks for pylon.
  * Fixed Acquisition method not setting FrameTriggerWait properly.
  * Changed limitbandwidth to a true/false option.
  * Implemented PylonC error macro to output line numbers where the error occurred.
  * Added orc_memcpy support.
  * Fixed camera reset code.
  * Removed support for chunk mode.
  * Updated documentation.

2017-02-17 Ingmars Melkis <zingmars@playgineering.com>

  * Release.

2016-08-26 Ingmars Melkis <zingmars@playgineering.com>

  * Created a simple image grabber plugin.

Copyright 2016-2018 Playgineering Ltd.
Copying and distribution of this file, with or without modification, are
permitted provided the copyright notice and this notice are preserved.


plugins/gstpylonsrc.h
/* GStreamer
 * Copyright (C) 2016-2017 Ingmars Melkis <zingmars@playgineering.com>
 * Copyright (C) 2018 Ingmars Melkis <contact@zingmars.me>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#ifndef _GST_PYLONSRC_H_
#define _GST_PYLONSRC_H_

#include <gst/base/gstpushsrc.h>
#include "pylonc/PylonC.h"

G_BEGIN_DECLS

#define GST_TYPE_PYLONSRC   (gst_pylonsrc_get_type())
#define GST_PYLONSRC(obj)   (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_PYLONSRC,GstPylonsrc))
#define GST_PYLONSRC_CLASS(klass)   (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_PYLONSRC,GstPylonsrcClass))
#define GST_IS_PYLONSRC(obj)   (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_PYLONSRC))
#define GST_IS_PYLONSRC_CLASS(obj)   (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_PYLONSRC))

typedef struct _GstPylonsrc GstPylonsrc;
typedef struct _GstPylonsrcClass GstPylonsrcClass;

struct _GstPylonsrc
{
  GstPushSrc base_pylonsrc;
  
  gint cameraId;
  PYLON_DEVICE_HANDLE deviceHandle; // Handle for the camera.
  PYLON_STREAMGRABBER_HANDLE streamGrabber; // Handler for camera's streams.
  PYLON_WAITOBJECT_HANDLE waitObject; // Handles timing out in the main loop.

  int32_t frameSize; // Size of a frame in bytes.
  int32_t payloadSize; // Size of a frame in bytes.
  guint64 frameNumber; // Fun note: At 120fps it will take around 4 billion years to overflow this variable.
  
  // Plugin parameters
  _Bool setFPS, continuousMode, limitBandwidth, demosaicing, centerx, centery, flipx, flipy;
  double fps, exposure, gain, blacklevel, gamma, balancered, balanceblue, balancegreen, redhue, redsaturation, yellowhue, yellowsaturation, greenhue, greensaturation, cyanhue, cyansaturation, bluehue, bluesaturation, magentahue, magentasaturation, sharpnessenhancement, noisereduction, autoexposureupperlimit, autoexposurelowerlimit, gainupperlimit, gainlowerlimit, brightnesstarget, transformation00, transformation01, transformation02, transformation10, transformation11, transformation12, transformation20, transformation21, transformation22;
  int64_t height, width, maxHeight, maxWidth, maxBandwidth, testImage, offsetx, offsety;
  gchar *imageFormat, *sensorMode, *lightsource, *autoexposure, *autowhitebalance, *autogain, *reset, *autoprofile, *transformationselector, *userid;
};

struct _GstPylonsrcClass
{
  GstPushSrcClass base_pylonsrc_class;
};

GType gst_pylonsrc_get_type (void);

G_END_DECLS

#endif


plugins/gstfpsfilter.h
/*
 * GStreamer fps filter element
 * Copyright (C) 2005 Thomas Vander Stichele <thomas@apestaart.org>
 * Copyright (C) 2005 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright (C) 2017 Ingmars Melkis <zingmars@playgineering.com>
 * Copyright (C) 2018 Ingmars Melkis <contact@zingmars.me>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef __GST_FPSFILTER_H__
#define __GST_FPSFILTER_H__

#include <gst/gst.h>

G_BEGIN_DECLS

/* #defines don't like whitespacey bits */
#define GST_TYPE_FPSFILTER \
  (gst_fps_filter_get_type())
#define GST_FPSFILTER(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_FPSFILTER,GstFpsFilter))
#define GST_FPSFILTER_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_FPSFILTER,GstFpsFilterClass))
#define GST_IS_FPSFILTER(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_FPSFILTER))
#define GST_IS_FPSFILTER_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_FPSFILTER))

typedef struct _GstFpsFilter      GstFpsFilter;
typedef struct _GstFpsFilterClass GstFpsFilterClass;

struct _GstFpsFilter
{
  GstElement element;
  guint64 frames, lastframetime, elapsedtime, reporttime;
  GstPad *sinkpad, *srcpad;
};

struct _GstFpsFilterClass 
{
  GstElementClass parent_class;
};

GType gst_fps_filter_get_type (void);

G_END_DECLS

#endif /* __GST_FPSFILTER_H__ */


plugins/gstfpsfilter.c
/* GStreamer
 * Copyright (C) 2017 Ingmars Melkis <zingmars@playgineering.com>
 * Copyright (C) 2018 Ingmars Melkis <contact@zingmars.me>
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/**
 * SECTION:element-fpsfilter
 *
 * A gstreamer element that calculates the FPS value of the stream.
 *
 * <refsect2>
 * <title>Example launch line</title>
 * |[
 * gst-launch-1.0  videotestsrc ! video/x-raw, framerate=1/1 ! videoconvert ! fpsfilter ! xvimagesink
 * ]|
 * </refsect2>
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <gst/gst.h>

#include "gstfpsfilter.h"

GST_DEBUG_CATEGORY_STATIC (gst_fps_filter_debug);
#define GST_CAT_DEFAULT gst_fps_filter_debug
#define GST_MESSAGE_OBJECT(obj, ...) GST_CAT_LEVEL_LOG(GST_CAT_DEFAULT, GST_LEVEL_NONE, obj, __VA_ARGS__)

/* Filter signals and args */
enum
{
  LAST_SIGNAL
};

/* Parameters */
enum
{
  PROP_0,
  PROP_REPORTTIME
};

/* the capabilities of the inputs and outputs.
 *
 * describe the real formats here.
 */
static GstStaticPadTemplate sink_factory = GST_STATIC_PAD_TEMPLATE ("sink",
    GST_PAD_SINK,
    GST_PAD_ALWAYS,
    GST_STATIC_CAPS ("ANY")
    );

static GstStaticPadTemplate src_factory = GST_STATIC_PAD_TEMPLATE ("src",
    GST_PAD_SRC,
    GST_PAD_ALWAYS,
    GST_STATIC_CAPS ("ANY")
    );

#define gst_fps_filter_parent_class parent_class
G_DEFINE_TYPE (GstFpsFilter, gst_fps_filter, GST_TYPE_ELEMENT);

static void gst_fps_filter_set_property (GObject * object, guint prop_id,
    const GValue * value, GParamSpec * pspec);
static void gst_fps_filter_get_property (GObject * object, guint prop_id,
    GValue * value, GParamSpec * pspec);

static gboolean gst_fps_filter_sink_event (GstPad * pad, GstObject * parent, GstEvent * event);
static GstFlowReturn gst_fps_filter_chain (GstPad * pad, GstObject * parent, GstBuffer * buf);

/* GObject vmethod implementations */

/* initialize the fpsfilter's class */
static void
gst_fps_filter_class_init (GstFpsFilterClass * klass)
{
  GObjectClass *gobject_class;
  GstElementClass *gstelement_class;

  gobject_class = (GObjectClass *) klass;
  gstelement_class = (GstElementClass *) klass;

  gobject_class->set_property = gst_fps_filter_set_property;
  gobject_class->get_property = gst_fps_filter_get_property;

  gst_element_class_set_details_simple(gstelement_class,
    "FPS counter",
    "Filter",
    "Calculates the time between frames and outputs the stream's framerate",
    "Ingmars Melkis <zingmars@playgineering.com>");

  gst_element_class_add_pad_template (gstelement_class,
      gst_static_pad_template_get (&src_factory));
  gst_element_class_add_pad_template (gstelement_class,
      gst_static_pad_template_get (&sink_factory));

  g_object_class_install_property(gobject_class, PROP_REPORTTIME, 
      g_param_spec_uint64("reporttime", "reporttime", "(Number) Time between fps reports in miliseconds (default - 1000)",
          0, G_MAXUINT64, 1000,
          (GParamFlags) (G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS)));
}

/* initialize the new element
 * instantiate pads and add them to element
 * set pad calback functions
 * initialize instance structure
 */
static void
gst_fps_filter_init (GstFpsFilter * filter)
{
  filter->sinkpad = gst_pad_new_from_static_template (&sink_factory, "sink");
  gst_pad_set_event_function (filter->sinkpad,
                              GST_DEBUG_FUNCPTR(gst_fps_filter_sink_event));
  gst_pad_set_chain_function (filter->sinkpad,
                              GST_DEBUG_FUNCPTR(gst_fps_filter_chain));
  GST_PAD_SET_PROXY_CAPS (filter->sinkpad);
  gst_element_add_pad (GST_ELEMENT (filter), filter->sinkpad);

  filter->srcpad = gst_pad_new_from_static_template (&src_factory, "src");
  GST_PAD_SET_PROXY_CAPS (filter->srcpad);
  gst_element_add_pad (GST_ELEMENT (filter), filter->srcpad);

  filter->frames = 0;
  filter->lastframetime = 0;
  filter->elapsedtime = 0;
  filter->reporttime = 1000;
}

static void
gst_fps_filter_set_property (GObject * object, guint prop_id,
    const GValue * value, GParamSpec * pspec)
{
  GstFpsFilter *filter = GST_FPSFILTER (object);

  switch (prop_id) {
    case PROP_REPORTTIME:
      filter->reporttime = g_value_get_uint64(value);
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
      break;
  }
}

static void
gst_fps_filter_get_property (GObject * object, guint prop_id,
    GValue * value, GParamSpec * pspec)
{
  GstFpsFilter *filter = GST_FPSFILTER (object);

  switch (prop_id) {
    case PROP_REPORTTIME:
      g_value_set_uint64(value, filter->reporttime);
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
      break;
  }
}

/* GstElement vmethod implementations */

/* this function handles sink events */
static gboolean
gst_fps_filter_sink_event (GstPad * pad, GstObject * parent, GstEvent * event)
{
  GstFpsFilter *filter;
  gboolean ret;

  filter = GST_FPSFILTER (parent);

  GST_LOG_OBJECT (filter, "Received %s event: %" GST_PTR_FORMAT,
      GST_EVENT_TYPE_NAME (event), event);

  switch (GST_EVENT_TYPE (event)) {
    case GST_EVENT_CAPS:
    {
      GstCaps * caps;

      gst_event_parse_caps (event, &caps);
      /* do something with the caps */

      /* and forward */
      ret = gst_pad_event_default (pad, parent, event);
      break;
    }
    default:
      ret = gst_pad_event_default (pad, parent, event);
      break;
  }
  return ret;
}

/* chain function
 * this function does the actual processing
 */
static GstFlowReturn
gst_fps_filter_chain (GstPad * pad, GstObject * parent, GstBuffer * buf)
{
  GstFpsFilter *filter;
  filter = GST_FPSFILTER (parent);

  if(GST_STATE(filter) == GST_STATE_PLAYING) {
    filter->frames = filter->frames + 1;
    #pragma GCC diagnostic ignored "-Wincompatible-pointer-types"
    gint64 time = GST_TIME_AS_MSECONDS(gst_clock_get_time(gst_element_get_clock(filter)));
    #pragma GCC diagnostic pop
    filter->elapsedtime = filter->elapsedtime + (time-filter->lastframetime);

    if(filter->elapsedtime >= filter->reporttime) {
      int64_t frames = filter->frames;
      int64_t elapsedtime = filter->elapsedtime;
      
      if(filter->elapsedtime != filter->reporttime) { //Most of the time our frame won't tick right on the dot. In those cases we're already processing a frame in the next cycle.
        frames=frames-1;
        elapsedtime = filter->elapsedtime-filter->elapsedtime%filter->reporttime;
      }

      GST_MESSAGE_OBJECT(filter, "FPS: %.0f (Calculated time per frame: %.1fms)", frames/((float)elapsedtime/1000), (float)elapsedtime/frames);

      filter->frames = filter->frames-frames;
      filter->elapsedtime = filter->elapsedtime-elapsedtime; //Carry over the extra time spent on the previous frame.
    }

    filter->lastframetime = time;
  }

  return gst_pad_push (filter->srcpad, buf);
}


/* entry point to initialize the plug-in
 * initialize the plug-in itself
 * register the element factories and other features
 */
static gboolean
fpsfilter_init (GstPlugin * fpsfilter)
{
  /* debug category for fltering log messages
   *
   * exchange the string 'Template fpsfilter' with your description
   */
  GST_DEBUG_CATEGORY_INIT (gst_fps_filter_debug, "fpsfilter",
      0, "Template fpsfilter");

  return gst_element_register (fpsfilter, "fpsfilter", GST_RANK_NONE,
      GST_TYPE_FPSFILTER);
}

/* PACKAGE: this is usually set by autotools depending on some _INIT macro
 * in configure.ac and then written into and defined in config.h, but we can
 * just set it ourselves here in case someone doesn't use autotools to
 * compile this code. GST_PLUGIN_DEFINE needs PACKAGE to be defined.
 */
#ifndef VERSION
#define VERSION "1.1.0"
#endif
#ifndef PACKAGE
#define PACKAGE "gstpylon"
#endif
#ifndef PACKAGE_NAME
#define PACKAGE_NAME "gstpylon"
#endif
#ifndef GST_PACKAGE_ORIGIN
#define GST_PACKAGE_ORIGIN "http://www.playgineering.com/"
#endif

/* gstreamer looks for this structure to register fpsfilters
 *
 * exchange the string 'Template fpsfilter' with your fpsfilter description
 */
GST_PLUGIN_DEFINE (
    GST_VERSION_MAJOR,
    GST_VERSION_MINOR,
    fpsfilter,
    "A plugin that calculates a stream's framerate value.",
    fpsfilter_init,
    VERSION,
    "LGPL",
    PACKAGE,
    GST_PACKAGE_ORIGIN
)
